### PR TITLE
Bingo v3.0

### PR DESCRIPTION
**Pull request by zfox23**
- New Bingo export tool (contact me for URL)
- Added a feature to the bouncer zone such that it no longer relies on hard-coded usernames
- Fixed a bug where a user could not get a card, enter the scanner, fail verification, then re-enter the scanner and get Bingo
- Added a "Bingo Winner" app, to be used by Bingo winners who win physical prizes

### CODE REVIEW

**Comment by AndySeattle** [comment]
passed; Error code: 4.

File: events/bingo/bingoWinnerApp/bingoWinner_app.js [hunk 1]
```diff
@@ -0,0 +1,181 @@
+/* eslint-disable no-magic-numbers */
```
Comment (robin-k-wilson) [inline events/bingo/bingoWinnerApp/bingoWinner_app.js]: why?

File: events/bingo/serverScripts/bingoWheel/bingoWheelServer.js [hunk 2]
```diff
@@ -31,13 +31,19 @@
     var currentRoundWinners = [];
     var BINGO_PRIZE_DOOR_1_TEXT = "{ff7674bb-5569-4381-b370-1dfa1d2a9723}";
     var BINGO_PRIZE_DOOR_1 = "{8724dc08-c0fb-4feb-bda1-8686023f8355}";
```
Comment (robin-k-wilson) [inline events/bingo/serverScripts/bingoWheel/bingoWheelServer.js]: Why are these hardcoded? I'm curious.

File: events/bingo/webApps/bingo.js [hunk 3]
```diff
@@ -348,16 +426,221 @@ function replaceCalledNumbers(newCalledNumbers, response) {
...
+        }
+    }
+
+    finalResponseHTML += `</html></body>`;
+
+    response.statusCode = 200;
+    response.setHeader('Content-Type', 'text/html');
+    return response.end(finalResponseHTML);
+}
+
```
Comment (robin-k-wilson) [inline events/bingo/webApps/bingo.js]: whitespace

File: events/bingo/bingoWinnerApp/bingoWinner_app.js [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+//
+// bingoWinnerApp.js
+// Created by Zach Fox on 2019-03-18
+// Copyright High Fidelity 2019
+//
+// Licensed under the Apache 2.0 License
+// See accompanying license file or http://apache.org/
+//
+
+/* globals */
```
[the remark this answers is not in the document]
Comment (zfox23) [reply events/bingo/bingoWinnerApp/bingoWinner_app.js]: nope! :D removing

File: events/bingo/serverScripts/bingoWheel/bingoWheelServer.js [hunk 5]
```diff
@@ -194,27 +200,31 @@
...
             var isAdding = data.isAdding;
 
-            function addOrRemove(array, username, isAdding) {
-                var idx = array.indexOf(username);
+            function addOrRemove(array, data, isAdding) {
+                var idx = array.indexOf(data);
 
                 if (idx > -1 && !isAdding) {
-                    console.log("Removing " + username + " from list of avatars currently in Prize Zone " + zoneNumber + "!");
+                    console.log("Removing " + data + " from list of avatars currently in Prize Zone " + zoneNumber + "!");
```
[the remark this answers is not in the document]
Comment (zfox23) [reply events/bingo/serverScripts/bingoWheel/bingoWheelServer.js]: This is one instance where it's really helpful to have a log!

File: events/bingo/serverScripts/bingoWheel/bingoWheelServer.js [hunk 6]
```diff
@@ -194,27 +200,31 @@
...
                     array.splice(idx, 1);
                 } else if (idx === -1 && isAdding) {
-                    console.log("Adding " + username + " to list of avatars currently in Prize Zone " + zoneNumber + "!");
-                    array.push(username);
+                    console.log("Adding " + data + " to list of avatars currently in Prize Zone " + zoneNumber + "!");
```
[the remark this answers is not in the document]
Comment (zfox23) [reply events/bingo/serverScripts/bingoWheel/bingoWheelServer.js]: This is one instance where it's really helpful to have a log!

File: events/bingo/bingoWinnerApp/bingoWinner_app.js (hunk 1 above)
Comment (zfox23) [reply events/bingo/bingoWinnerApp/bingoWinner_app.js]: cuz i copied and pasted lazily :D removing

File: events/bingo/serverScripts/bingoWheel/bingoWheelServer.js (hunk 2 above)
Comment (zfox23) [reply events/bingo/serverScripts/bingoWheel/bingoWheelServer.js]: they have to be for Bingo to work - there's no other way I can think of that would ensure that our content works reliably.

File: events/bingo/webApps/bingo.js (hunk 3 above)
Comment (zfox23) [reply events/bingo/webApps/bingo.js]: rats! got it

**Pull request merged**